File: app/src/main/java/ua/meugen/android/pdfcomposer/ui/activities/createnew/fragments/createnew/presenter/CreateNewPresenterImpl.java
package ua.meugen.android.pdfcomposer.ui.activities.createnew.fragments.createnew.presenter;

import java.util.ArrayList;
import java.util.List;

import ua.meugen.android.pdfcomposer.model.PdfExporter;
import ua.meugen.android.pdfcomposer.model.data.PageContent;
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.presenter.BasePresenter;
import ua.meugen.android.pdfcomposer.ui.activities.createnew.fragments.createnew.state.CreateNewState;
import ua.meugen.android.pdfcomposer.ui.activities.createnew.fragments.createnew.view.CreateNewView;


public class CreateNewPresenterImpl extends BasePresenter<CreateNewState>
        implements CreateNewPresenter {

    private final CreateNewView view;
    private final PdfExporter pdfExporter;

    private List<PageContent> pages;

    public CreateNewPresenterImpl(
            final CreateNewView view,
            final PdfExporter pdfExporter) {
        this.view = view;
        this.pdfExporter = pdfExporter;
    }

    @Override
    public void onSaveState(final CreateNewState state) {
        super.onSaveState(state);
        state.setPages(pages);
    }

    @Override
    public void onRestoreState(final CreateNewState state) {
        super.onRestoreState(state);
        this.pages = state.getPages();
    }

    private PageContent getLastPage() {
        return pages.get(pages.size() - 1);
    }

    @Override
    public void setup() {
        if (pages == null) {
            pages = new ArrayList<>();
            pages.add(new PageContent());
        }
        final PageContent lastPage = getLastPage();
        view.displayContent(lastPage.getImageUri(), lastPage.getText());
    }

    @Override
    public void onImageSelected(final String imageUri) {
        final PageContent lastPage = getLastPage();
        lastPage.setImageUri(imageUri);
        view.displayContent(imageUri, lastPage.getText());
    }

    @Override
    public void onTextChanged(final String text) {
        final PageContent lastPage = getLastPage();
        lastPage.setText(text);
    }

    private boolean checkLastPageValid() {
        final PageContent lastPage = getLastPage();
        return lastPage.getImageUri() != null
                && lastPage.getText() != null
                && !lastPage.getText().isEmpty();
    }

    @Override
    public void addPage() {
        if (!checkLastPageValid()) {
            view.pageDidntComplete();
            return;
        }
        pages.add(new PageContent());
        view.displayContent(null, null);
    }

    @Override
    public void exportPdf() {
        if (!checkLastPageValid()) {
            view.pageDidntComplete();
            return;
        }
        pdfExporter.exportPages(pages);
    }
}

File: app/src/main/java/ua/meugen/android/pdfcomposer/ui/activities/createnew/fragments/createnew/presenter/CreateNewPresenterImpl.kt
package ua.meugen.android.pdfcomposer.ui.activities.createnew.fragments.createnew.presenter

import java.util.ArrayList

import ua.meugen.android.pdfcomposer.model.PdfExporter
import ua.meugen.android.pdfcomposer.model.data.PageContent
import ua.meugen.android.pdfcomposer.model.utils.toArrayList
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.presenter.BasePresenter
import ua.meugen.android.pdfcomposer.ui.activities.createnew.fragments.createnew.state.CreateNewState
import ua.meugen.android.pdfcomposer.ui.activities.createnew.fragments.createnew.view.CreateNewView


class CreateNewPresenterImpl(
        private val view: CreateNewView,
        private val pdfExporter: PdfExporter) : BasePresenter<CreateNewState>(), CreateNewPresenter {

    private var pages: MutableList<PageContent>? = null

    private val lastPage: PageContent
        get() = pages!![pages!!.size - 1]

    override fun onSaveState(state: CreateNewState) {
        super.onSaveState(state)
        state.pages = pages
    }

    override fun onRestoreState(state: CreateNewState) {
        super.onRestoreState(state)
        this.pages = state.pages?.toArrayList
    }

    override fun setup() {
        if (pages == null) {
            pages = ArrayList()
            pages!!.add(PageContent())
        }
        val lastPage = lastPage
        view.displayContent(lastPage.imageUri, lastPage.text)
    }

    override fun onImageSelected(imageUri: String) {
        val lastPage = lastPage
        lastPage.imageUri = imageUri
        view.displayContent(imageUri, lastPage.text!!)
    }

    override fun onTextChanged(text: String) {
        val lastPage = lastPage
        lastPage.text = text
    }

    private fun checkLastPageValid(): Boolean {
        val lastPage = lastPage
        return (lastPage.imageUri != null
                && lastPage.text != null
                && !lastPage.text!!.isEmpty())
    }

    override fun addPage() {
        if (!checkLastPageValid()) {
            view.pageDidntComplete()
            return
        }
        pages!!.add(PageContent())
        view.displayContent(null, null)
    }

    override fun exportPdf() {
        if (!checkLastPageValid()) {
            view.pageDidntComplete()
            return
        }
        pdfExporter.exportPages(pages!!)
    }
}
